hummingbot/connector/exchange/bitmax/bitmax_order_book_tracker.py

```diff
@@ -0,0 +1,109 @@
+
+    @property
+    def exchange_name(self) -> str:
+        """
+        Name of the current exchange
+        """
+        return constants.EXCHANGE_NAME
+
+    async def _track_single_book(self, trading_pair: str):
+        """
+        Update an order book with changes from the latest batch of received messages
+        """
+        past_diffs_window: Deque[BitmaxOrderBookMessage] = deque()
+        self._past_diffs_windows[trading_pair] = past_diffs_window
+
+        message_queue: asyncio.Queue = self._tracking_message_queues[trading_pair]
+        order_book: BitmaxOrderBook = self._order_books[trading_pair]
+        active_order_tracker: BitmaxActiveOrderTracker = self._active_order_trackers[trading_pair]
+
+        last_message_timestamp: float = time.time()
+        diff_messages_accepted: int = 0
+
+        while True:
+            try:
+                message: BitmaxOrderBookMessage = None
+                saved_messages: Deque[BitmaxOrderBookMessage] = self._saved_message_queues[trading_pair]
+                # Process saved messages first if there are any
+                if len(saved_messages) > 0:
+                    message = saved_messages.popleft()
+                else:
+                    message = await message_queue.get()
+
+                if message.type is OrderBookMessageType.DIFF:
+                    bids, asks = active_order_tracker.convert_diff_message_to_order_book_row(message)
+                    order_book.apply_diffs(bids, asks, message.update_id)
+                    past_diffs_window.append(message)
+                    while len(past_diffs_window) > self.PAST_DIFF_WINDOW_SIZE:
+                        past_diffs_window.popleft()
+                    diff_messages_accepted += 1
+
+                    # Output some statistics periodically.
+                    now: float = time.time()
+                    if int(now / 60.0) > int(last_message_timestamp / 60.0):
+                        self.logger().debug("Processed %d order book diffs for %s.",
+                                            diff_messages_accepted, trading_pair)
+                        diff_messages_accepted = 0
+                    last_message_timestamp = now
+                elif message.type is OrderBookMessageType.SNAPSHOT:
+                    past_diffs: List[BitmaxOrderBookMessage] = list(past_diffs_window)
+                    # only replay diffs later than snapshot, first update active order with snapshot then replay diffs
+                    replay_position = bisect.bisect_right(past_diffs, message)
+                    replay_diffs = past_diffs[replay_position:]
+                    s_bids, s_asks = active_order_tracker.convert_snapshot_message_to_order_book_row(message)
+                    order_book.apply_snapshot(s_bids, s_asks, message.update_id)
+                    for diff_message in replay_diffs:
+                        d_bids, d_asks = active_order_tracker.convert_diff_message_to_order_book_row(diff_message)
```

Comment:
As mentioned, `convert_snapshot_message_to_order_book_row` and `convert_diff_message_to_order_book_row` should not be in `bitmax_utils.py`

Reply:
lets track this over at https://github.com/CoinAlpha/hummingbot/pull/2823#discussion_r563436683